python-frontend/src/test/java/org/sonar/python/tree/CallExpressionImplTest.java

```diff
@@ -102,4 +111,95 @@ public void not_callable_callee_symbol_type() {
   public void null_callee_symbol_type() {
     assertThat(lastExpression("x()").type()).isEqualTo(InferredTypes.anyType());
   }
+
+  @Test
+  public void method_call_with_declared_type() {
+    assertThat(lastExpression(
+      "def foo(param: str):",
+      "  x = param.capitalize()",
+      "  x"
+    ).type()).isEqualTo(DECL_STR);
+
+    assertThat(lastExpression(
+      "def foo(param: A):",
+      "  x = param.capitalize()",
+      "  x"
+    ).type()).isEqualTo(anyType());
+
+    assertThat(lastExpression(
+      "def foo(param):",
+      "  x = param.capitalize()",
+      "  x"
+    ).type()).isEqualTo(anyType());
+
+    // TODO: handle member access dependencies for declared type
+    assertThat(lastExpression(
+      "def foo(param: str):",
+      "  x = param.capitalize().capitalize()",
+      "  x"
+    ).type()).isEqualTo(anyType());
+
+    assertThat(lastExpression(
+      "class A:",
+      "  def meth() -> int: ...",
+      "def foo(param: A):",
+      "  x = param.meth()",
+      "  x"
+    ).type()).isEqualTo(DECL_INT);
+  }
+
+  @Test
+  public void method_call_with_declared_type_union_and_optional() {
+    ClassSymbolImpl union = new ClassSymbolImpl("Union", "typing.Union");
+    List<DeclaredType> typeArgs = Stream.of(DECL_INT, DECL_STR).map(DeclaredType.class::cast).collect(Collectors.toList());
+    assertThat(lastExpression(
+      "from typing import Union",
+      "class A:",
+      "  def meth() -> Union[int, str]: ...",
+      "def foo(param: A):",
+      "  x = param.meth()",
+      "  x"
+    ).type()).isEqualTo(new DeclaredType(union, typeArgs));
+
+    // TODO: handle UnionType of annotation coming from typeshed
```

Comment:
Should we create & reference a ticket for that here?

Reply:
done: https://jira.sonarsource.com/browse/SONARPY-782